tree/tree/src/TTree.cxx

```diff
@@ -4492,62 +4507,64 @@ Int_t TTree::Fill()
-               fAutoSave = fAutoFlush*(fAutoSave/fAutoFlush);
+            } else if (fAutoSave > 0) {
+               fAutoSave = fAutoFlush * (fAutoSave / fAutoFlush);
             }
-            if (fAutoSave!=0 && fEntries >= fAutoSave) AutoSave();    // FlushBaskets not called in AutoSave
-            if (gDebug > 0) Info("TTree::Fill","First AutoFlush.  fAutoFlush = %lld, fAutoSave = %lld\n", fAutoFlush, fAutoSave);
-         }
-      } else if (fNClusterRange && fAutoFlush && ( (fEntries-fClusterRangeEnd[fNClusterRange-1]) % fAutoFlush == 0)  ) {
-         if (fAutoSave != 0 && fEntries%fAutoSave == 0) {
-            //We are at an AutoSave point. AutoSave flushes baskets and saves the Tree header
-            AutoSave("flushbaskets");
-            if (gDebug > 0) Info("TTree::Fill","AutoSave called at entry %lld, fZipBytes=%lld, fSavedBytes=%lld\n",fEntries,GetZipBytes(),fSavedBytes);
-         } else {
-            //We only FlushBaskets
-            FlushBaskets();
-            if (gDebug > 0) Info("TTree::Fill","FlushBasket called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n",fEntries,GetZipBytes(),fFlushedBytes);
-         }
-         fFlushedBytes = GetZipBytes();
-      } else if (fNClusterRange == 0 && fEntries > 1 && fAutoFlush && fEntries%fAutoFlush == 0) {
-         if (fAutoSave != 0 && fEntries%fAutoSave == 0) {
-            //We are at an AutoSave point. AutoSave flushes baskets and saves the Tree header
-            AutoSave("flushbaskets");
-            if (gDebug > 0) Info("TTree::Fill","AutoSave called at entry %lld, fZipBytes=%lld, fSavedBytes=%lld\n",fEntries,GetZipBytes(),fSavedBytes);
-         } else {
-            //We only FlushBaskets
-            FlushBaskets();
-            if (gDebug > 0) Info("TTree::Fill","FlushBasket called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n",fEntries,GetZipBytes(),fFlushedBytes);
+
+            if (fAutoSave != 0 && fEntries >= fAutoSave)
+               autoSave = true;
+
+            if (gDebug > 0)
+               Info("TTree::Fill", "First AutoFlush.  fAutoFlush = %lld, fAutoSave = %lld\n", fAutoFlush, fAutoSave);
          }
-         fFlushedBytes = GetZipBytes();
+      } else {
+         // Check if we need to auto flush
+         if (fAutoFlush) {
+            if (fNClusterRange == 0)
+               autoFlush = fEntries > 1 && fEntries % fAutoFlush == 0;
+            else
+               autoFlush = (fEntries - fClusterRangeEnd[fNClusterRange - 1]) % fAutoFlush == 0;
+         }
+         // Check if we need to auto save
+         if (fAutoSave)
+            autoSave = fEntries % fAutoSave == 0;
       }
    }
+
+   if (autoFlush) {
+      FlushBaskets();
+      if (gDebug > 0)
+         Info("TTree::Fill", "FlushBaskets() called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n", fEntries,
+              GetZipBytes(), fFlushedBytes);
+      fFlushedBytes = GetZipBytes();
+   }
+
+   if (autoSave) {
+      AutoSave();
```

Comment:
One of the previous incarnation add a comment here that helped (i.e. something about this call to autosave not calling AutoFlush)

Reply:
Ah, sure. I will add it back.

Reply:
I added the comment back and pushed again. Will merge once Jenkins is done.